File: hw/astudillo/U3/HW16Maze/maze_python/src/maze/controller/room.py
from typing import Dict
from ..model.coordinate import Coordinate
from ..model.wall import Wall


class Room:
    def __init__(self, coordinate: Coordinate):
        self._coordinate = coordinate
        self._walls: Dict[str, Wall] = {
            "UP": Wall(False),
            "DOWN": Wall(False),
            "LEFT": Wall(False),
            "RIGHT": Wall(False)
        }
        self._is_entrance_room = False
        self._is_exit_room = False

    @property
    def coordinate(self) -> Coordinate:
        return self._coordinate

    def open_wall(self, direction: str):
        if direction in self._walls:
            self._walls[direction].open()

    def close_wall(self, direction: str):
        if direction in self._walls:
            self._walls[direction].close()

    def has_door(self, direction: str) -> bool:
        return direction in self._walls and self._walls[direction].is_open

    def has_wall(self, direction: str) -> bool:
        return not self.has_door(direction)

    @property
    def is_entrance(self) -> bool:
        return self._is_entrance_room

    @is_entrance.setter
    def is_entrance(self, value: bool):
        self._is_entrance_room = value

    @property
    def is_exit(self) -> bool:
        return self._is_exit_room

    @is_exit.setter
    def is_exit(self, value: bool):
        self._is_exit_room = value

    @property
    def display_symbol(self) -> str:
        if self.is_entrance:
            return "S"
        elif self.is_exit:
            return "E"
        else:
            return " "

    def __str__(self) -> str:
        return f"Room at {self._coordinate} - {self.display_symbol}"

File: hw/astudillo/U3/HW16Maze/maze/src/main/java/org/example/edu/ec/espe/maze/controller/Room.java
package org.example.edu.ec.espe.maze.controller;

import org.example.edu.ec.espe.maze.model.Coordinate;
import org.example.edu.ec.espe.maze.model.Wall;
import java.util.HashMap;
import java.util.Map;

public class Room {
    private final Coordinate coordinate;
    private final Map<String, Wall> walls;
    private boolean isEntranceRoom;
    private boolean isExitRoom;

    public Room(Coordinate coordinate) {
        this.coordinate = coordinate;
        this.walls = new HashMap<>();
        this.isEntranceRoom = false;
        this.isExitRoom = false;
        walls.put("UP", new Wall(false));
        walls.put("DOWN", new Wall(false));
        walls.put("LEFT", new Wall(false));
        walls.put("RIGHT", new Wall(false));
    }

    public Coordinate getCoordinate() { return coordinate; }
    public void openWall(String direction) { if (walls.containsKey(direction)) walls.get(direction).setOpen(true); }
    public void closeWall(String direction) { if (walls.containsKey(direction)) walls.get(direction).setOpen(false); }
    public boolean hasDoor(String direction) { return walls.containsKey(direction) && walls.get(direction).isOpen(); }
    public boolean isEntrance() { return isEntranceRoom; }
    public void setEntrance(boolean isEntranceRoom) { this.isEntranceRoom = isEntranceRoom; }
    public boolean isExit() { return isExitRoom; }
    public void setExit(boolean isExitRoom) { this.isExitRoom = isExitRoom; }
    public String getDisplaySymbol() { return isEntrance() ? "S" : isExit() ? "E" : " "; }
    public boolean hasWall(String direction) { return !hasDoor(direction); }
}
